src/Quantum/Quantum.jl

```diff
@@ -8,73 +9,84 @@ Tensor Network [`Ansatz`](@ref) that has a notion of sites and directionality (i
+sites(tn::TensorNetwork, ::Val{:out}) = first.(filter(==(:out) ∘ last, keys(tn[:plug])))
+
+labels(tn::TensorNetwork, ::Val{:plug}) = unique(values(tn[:plug]))
+labels(tn::TensorNetwork, ::Val{:plug}, site) = labels(tn, Val(:in), site) ∪ labels(tn, Val(:out), site)
+labels(tn::TensorNetwork, ::Val{:in}) = map(last, Iterators.filter((((_, dir), _),) -> dir === :in, tn[:plug]))
+labels(tn::TensorNetwork, ::Val{:in}, site) = tn[:plug][(site, :in)]
+labels(tn::TensorNetwork, ::Val{:out}) = map(last, Iterators.filter((((_, dir), _),) -> dir === :out, tn[:plug]))
+labels(tn::TensorNetwork, ::Val{:out}, site) = tn[:plug][(site, :out)]
+labels(tn::TensorNetwork, ::Val{:virtual}) = setdiff(labels(tn, Val(:all)), labels(tn, Val(:plug)))
+
+tensors(tn::TensorNetwork{<:Quantum}, site::Integer, args...) = tensors(plug(tn), tn, site, args...)
+tensors(::Type{State}, tn::TensorNetwork{<:Quantum}, site) = select(tn, labels(tn, :out, site)) |> only
+@valsplit 4 tensors(T::Type{Operator}, tn::TensorNetwork{<:Quantum}, site, dir::Symbol) =
+    throw(MethodError(sites, "dir=$dir not recognized"))
+tensors(T::Type{Operator}, tn::TensorNetwork{<:Quantum}, site, ::Val{:in}) = select(tn, labels(tn, :in, site)) |> only
+tensors(T::Type{Operator}, tn::TensorNetwork{<:Quantum}, site, ::Val{:out}) = select(tn, labels(tn, :out, site)) |> only
 
 function Base.hcat(A::TensorNetwork{QA}, B::TensorNetwork{QB}) where {QA<:Quantum,QB<:Quantum}
-    outsites(A) != insites(B) && throw(DimensionMismatch("insites(B) must be equal to outsites(A) to connect them"))
+    sites(A, :out) != sites(B, :in) &&
+        throw(DimensionMismatch("sites(B,:in) must be equal to sites(A,:out) to connect them"))
 
     # rename connector indices
-    newinds = Dict([s => Symbol(uuid4()) for s in outsites(A)])
+    newinds = Dict([s => Symbol(uuid4()) for s in sites(A, :out)])
 
-    A = replace(A, [nameof(i) => newinds[site(i)] for i in outsiteinds(A)]...)
-    B = replace(B, [nameof(i) => newinds[site(i)] for i in insiteinds(B)]...)
+    A = replace(A, [labels(A, :out, site) => newinds[site] for site in sites(A, :out)]...)
+    B = replace(B, [labels(B, :in, site) => newinds[site] for site in sites(B, :in)]...)
 
     # remove plug metadata on connector indices
-    for i in values(newinds)
-        delete!(A.inds[i].meta, :plug)
-        delete!(B.inds[i].meta, :plug)
+    for site in sites(A, :out)
+        delete!(A[:plug], (site, :out))
+    end
+    for site in sites(B, :in)
+        delete!(B[:plug], (site, :in))
     end
 
     # rename inner indices of B to avoid hyperindices
-    replace!(B, [nameof(i) => Symbol(uuid4()) for i in innerinds(B)]...)
+    replace!(B, [i => Symbol(uuid4()) for i in labels(B, :inner)]...)
 
     # merge tensors and indices
-    tn = TensorNetwork{Tuple{QA,QB}}(; merge(A.meta, B.meta)...)
-    append!(tn, A)
-    append!(tn, B)
+    tn = TensorNetwork{Tuple{QA,QB}}([tensors(A)..., tensors(B)...]; mergewith(merge, A.metadata, B.metadata)...)
```

Comment:
```suggestion
    tn = TensorNetwork{Tuple{QA,QB}}([tensors(A)..., tensors(B)...]; merge(A.metadata, B.metadata)...)
```
`mergewith` gives error, use legacy `merge` instead.

Reply:
But we need `mergewith` in order to `merge` `Dict`s inside `A.metadata` and `B.metadata`.

Can you show me the error?

Reply:
For example:
```julia
julia> using Tenet
[ Info: Precompiling Tenet [85d41934-b9cd-44e1-8730-56d86f15f3ec]

julia> ψ = rand(MatrixProduct{State,Open}, n = 4, p = 2, χ = 32)
TensorNetwork{MatrixProduct{State, Open}}(#tensors=4, #labels=7)

julia> ψ[:plug]
Dict{Tuple{Int64, Symbol}, Symbol} with 4 entries:
  (2, :out) => Symbol("cc1fe63e-724c-4541-9140-294e30b8577f")
  (3, :out) => Symbol("2e180458-edce-4ce1-9ecd-c11765fc185b")
  (1, :out) => Symbol("439d438b-53e8-4217-9a23-1f839e89ea42")
  (4, :out) => Symbol("67de3b7b-2c47-4439-aba5-93fc5be6b374")

julia> ψ'[:plug]
Dict{Tuple{Int64, Symbol}, Symbol} with 4 entries:
  (4, :in) => Symbol("67de3b7b-2c47-4439-aba5-93fc5be6b374")
  (2, :in) => Symbol("cc1fe63e-724c-4541-9140-294e30b8577f")
  (3, :in) => Symbol("2e180458-edce-4ce1-9ecd-c11765fc185b")
  (1, :in) => Symbol("439d438b-53e8-4217-9a23-1f839e89ea42")

julia> mergewith(merge, ψ.metadata, ψ'.metadata)
ERROR: MethodError: no method matching merge(::Int64, ::Int64)
Closest candidates are:
  merge(::NamedTuple, ::Any) at namedtuple.jl:299
Stacktrace:
 [1] mergewith!(combine::typeof(merge), d1::Dict{Symbol, Any}, d2::Dict{Symbol, Any})
   @ Base ./dict.jl:752
 [2] mergewith(combine::Function, d::Dict{Symbol, Any}, others::Dict{Symbol, Any})
   @ Base ./abstractdict.jl:392
 [3] top-level scope
   @ REPL[6]:1
```

Reply:
@mofeing What about this? The problem is not solved

Reply:
Yep, I fixed yesterday but forgot to submit. Pushed now.

Reply:
Anyway, I think this will probably be better solved when I introduce "Layered Graphs" like functionality into `TensorNetwork`, which probably will be between versions 0.2 (the one to be released after this PR) and 0.3.